### ISSUE TITLE
Few shot learning

### ISSUE THREAD

**Issue by AkhilBhavanam** [ultralytics/yolov5#1750]
## ❔Question
Suggest few good changes that can be made to yolov5 to do few shot learning ?
## Additional context
Approaches tried : freezing all the layers except the last and trained the model with k samples from each class by using reduced learning rate and  using yolov5x coco weights ?

**Comment by glenn-jocher** [comment]
@AkhilBhavanam I think 'few shot learning' is a classification (resnet50) concept that's not been proposed for detection tasks like YOLOv5. Please send references if I'm incorrect.

**Comment by AkhilBhavanam** [comment]
@glenn-jocher  Yeah, but recently there is a paper which used the few shot learning for object detection as well, they have used faster rcnn to perform the tasks, they have also tried yolov3 I guess, So I am thinking to do this few shot learning using yolov5. Here are the references 
https://arxiv.org/abs/2003.06957
https://github.com/ucbdrive/few-shot-object-detection

**Comment by glenn-jocher** [comment]
@AkhilBhavanam ah interesting, thanks for the link. I see you tried some experiments yourself. In general there are a variety of limitations to the current setup that few shot and active learning may be able to help, but the execution risk is high, in that you may invest a lot of time and resources with little potential return, so its not something we can really explore ourselves as much as we'd like due to our ongoing maintenance commitments to the repo, which are using up most of our time.

In general if I were to approach this task, I think one key might be to disassociate the detection from the classification tasks. In YOLOv5 an object is detected, regressed and classified in parallel, and I don't see any reason that few shot learning should not be able to use the same regression outputs with minimal changes, so you might freeze the four regression outputs but leave the confidence output and classification outputs unfrozen to learn new examples. 

And then of course you could massively recycle your new objects with a high augmentation policy, i.e. if you have a new object, i.e. 'face' you want to add to COCO, you could gather 10 images with faces, and add a very severe augmentation to train these thousands of times each. 

One issue here is that I see significant interest for 'adding a class' to COCO. i.e. you like the COCO classes but you want to detect faces also, then people expect to simply train on faces while retaining the previous information, which is not possible at the moment, so forgetting previous knowledge is a common issue preventing extension of vision ML to new scenarios as well. It's all interesting stuff.

Any contributions you can make to the field would be much appreciated!

**Comment by glenn-jocher** [comment]
I'll add this slide here as well, since it's relevant. You can still train on infinite representations of the same underlying 'few' images, no matter how few they are.

<img width="1057" alt="Screen Shot 2020-12-21 at 10 42 35 AM" src="https://user-images.githubusercontent.com/26833433/102810827-443ec400-4379-11eb-86ff-32b0d7851049.png">

**Comment by AkhilBhavanam** [comment]
@glenn-jocher
Thanks for such beautiful idea 'freezing regression outputs' , as in YOLOV5 all are done in parallel, can we freeze only regression outputs? If yes, how?

And that's true that the model forgets the previous information, to avoid this one idea is to use 'k' samples from each class( both old and new classes) and train the network, instead of only training with the new classes.

**Comment by glenn-jocher** [comment]
@AkhilBhavanam yes, k samples from each class is not a bad idea, where k is similar to the new class counts I suppose. 

There is the class imbalance problem with new classes, though imbalances seem to not be too bad of an issue when the model is trained sufficiently. COCO has severe class imbalance, and last time I checked mAP per class did not have much correlation with class counts in the training data.

I wonder if there is some simpler way to handle forgetting. If training included some additional mechanism for simply maintaining the strength of existing outputs (without requiring the original training data), while simultaneously optimizing new outputs, that would be very interesting.

**Comment by glenn-jocher** [comment]
> I wonder if there is some simpler way to handle forgetting. If training included some additional mechanism for simply maintaining the strength of existing outputs (without requiring the original training data), while simultaneously optimizing new outputs, that would be very interesting.

Perhaps if training took a 2-step approach:
- forward 1, save output
- loss, backward, optimize

additional step to retain existing class behavior:
- loss2 = (forward 2 - forward 1)_existing_classes_only, backward2, optimze2

**Comment by glenn-jocher** [comment]
@AkhilBhavanam but to answer your question, I just remembered that you can't set individual elements of a tensor to require_gradient False, you can only change a tensor as a whole, so the idea to freeze the regression outputs is not possible.

**Comment by AkhilBhavanam** [comment]
@glenn-jocher 
2-step approach seems to be good, but here we need to separate out the exisiting classes forward2 right? I agree that freezing the regression outputs is not possible in case of YOLO.
And I am thinking of another approach to use 'k' samples from all classes and severely augment them and train the model, and to severely augment them can I get few suggestions for the values of hyper parameters.(let us suppose we have 10 images of all the classes), If we want model to generalize well, It would be great if you can suggest the values of hyper parameters.

Instead of training the whole network, can we expect the same performance if we freeze the backbone and perorm the same task as above?

To briefly explain my task, I am trying to add 'Mask', 'Door', 'Sanitizer' and few more classes to the COCO classes

**Comment by nmntz** [comment]
Hello @glenn-jocher! Is there a notebook that showcases few shot learning with YOLOv5?
